x-pack/filebeat/input/azureblobstorage/job.go

```diff
@@ -114,38 +108,120 @@ func (j *job) extractData(ctx context.Context) (*bytes.Buffer, error) {
 		}
 	}()
 
-	_, err = downloadedData.ReadFrom(reader)
+	updatedReader, err := j.addGzipDecoderIfNeeded(reader)
+	err = j.readJsonAndPublish(ctx, updatedReader, id)
 	if err != nil {
-		return nil, fmt.Errorf("failed to read data from blob with error: %w", err)
+		return fmt.Errorf("failed to read data from blob with error: %w", err)
 	}
 
-	return downloadedData, err
+	return err
 }
 
-func (j *job) createEventFields(message string) mapstr.M {
-	return mapstr.M{
-		"message": message, // original stringified data
-		"log": mapstr.M{
-			"file": mapstr.M{
-				"path": j.blobURL,
-			},
-		},
-		"azure": mapstr.M{
-			"storage": mapstr.M{
-				"container": mapstr.M{
-					"name": j.src.ContainerName,
+func (j *job) readJsonAndPublish(ctx context.Context, r io.Reader, id string) error {
+	dec := json.NewDecoder(r)
+	dec.UseNumber()
+	var offset int64
+	for dec.More() && ctx.Err() == nil {
+		var item json.RawMessage
+		offset = dec.InputOffset()
```

Comment:
I've removed the offset from the client while downloading the blobs.. and am manually seeking it via the decoding loop. This is allowing for proper tracking of gzipped file streams since they have been already decoded after the being downloaded. This also works for unzipped files but is a little less efficient .. but I feel the trade off is good in terms of the functionality.

Reply:
There is the option of keeping the offsets if you are not working with compressed data.

Reply:
Have improved the offset tracking with conditional logic @efd6